operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/managed/ManagedDependentResourceContext.java

```diff
@@ -70,43 +68,59 @@ public Optional put(Object key, Object value) {
+  public ManagedDependentResourceContext(Map<String, DependentResource> dependentResources) {
+    this.dependentResources = dependentResources;
   }
 
   /**
    * Retrieve all the known {@link DependentResource} implementations
    *
-   * @return a list of known {@link DependentResource} implementations
+   * @return known {@link DependentResource} implementations keyed by their name
    */
-  public List<DependentResource> getDependentResources() {
+  public Map<String, DependentResource> getDependentResources() {
     return dependentResources;
   }
 
   /**
    * Retrieve the dependent resource implementation associated with the specified resource type.
    *
-   * @param resourceClass the dependent resource class for which we want to retrieve the associated
-   *        dependent resource implementation
+   * @param name the name of the {@link DependentResource} implementation we want to retrieve
+   * @param resourceClass the resource class for which we want to retrieve the associated dependent
+   *        resource implementation
    * @param <T> the type of the resources for which we want to retrieve the associated dependent
    *        resource implementation
    * @return the associated dependent resource implementation if it exists or an exception if it
    *         doesn't or several implementations are associated with the specified resource type
    */
-  @SuppressWarnings("unchecked")
-  public <T extends DependentResource> T getDependentResource(Class<T> resourceClass) {
-    var resourceList =
-        dependentResources.stream()
-            .filter(dr -> dr.getClass().equals(resourceClass))
-            .collect(Collectors.toList());
-    if (resourceList.isEmpty()) {
+  @SuppressWarnings({"unchecked"})
+  public <T> DependentResource<T, ?> getDependentResource(String name, Class<T> resourceClass) {
+    DependentResource dependentResource = getDependentResource(name);
+    final var actual = dependentResource.resourceType();
+    if (!actual.equals(resourceClass)) {
       throw new OperatorException(
-          "No dependent resource found for class: " + resourceClass.getName());
+          "Dependent resource implementation doesn't match expected resource type, was: "
+              + actual.getName() + " instead of: " + resourceClass.getName());
     }
-    if (resourceList.size() > 1) {
-      throw new OperatorException(
-          "More than one dependent resource found for class: " + resourceClass.getName());
+    return dependentResource;
+  }
+
+  private DependentResource getDependentResource(String name) {
+    var dependentResource = dependentResources.get(name);
+    if (dependentResource == null) {
+      throw new OperatorException("No dependent resource found with name: " + name);
+    }
+    return dependentResource;
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T extends DependentResource> T getAdaptedDependentResource(String name,
```

Comment:
I guess it is from cdi proxy. Why should we make a difference now that the key is the name ?

Reply:
No, this actually has nothing to do with CDI proxies, just a convenience method to retrieve the dependent resource using its actual class instead of the more generic `DependentResource` interface.

Reply:
I don't get the diff with `public <T> DependentResource<T, ?> getDependentResource(String name, Class<T> resourceClass)`.
`getAdaptedDependentResource` can be used instead of `getDependentResource`.

Reply:
Yes, not sure which one will be kept :) `getDependentResource` is more generic so could be useful in more "automated" scenarios where knowing the actual type of the dependent resource is not practical. Basically, if you are using managed dependent resources, it kind of defeats the purpose to retrieve one implementation using its type. I see that as a failure in the API design.